lms/services/roster.py

```diff
@@ -373,21 +504,105 @@ def _get_roster_roles(self, roster) -> list[LTIRole]:
         return self._lti_role_service.get_roles(list(roles))
 
-    def _get_lti_registration(self, lms_course) -> LTIRegistration:
-        lti_registration = self._db.scalars(
-            select(LTIRegistration)
-            .join(ApplicationInstance)
-            .where(LMSCourseApplicationInstance.lms_course_id == lms_course.id)
+    def _get_application_instance(self, lms_course) -> ApplicationInstance:
+        return self._db.scalars(
+            select(ApplicationInstance)
             .join(LMSCourseApplicationInstance)
-            .order_by(LTIRegistration.updated.desc())
+            .where(LMSCourseApplicationInstance.lms_course_id == lms_course.id)
+            .order_by(LMSCourseApplicationInstance.updated.desc())
         ).first()
-        assert lti_registration, "No LTI registration found for LMSCourse."
-        return lti_registration
+
+    def _get_lti_registration(self, lms_course) -> LTIRegistration:
+        ai = self._get_application_instance(lms_course)
+        assert ai.lti_registration, "No LTI registration found for LMSCourse."
+        return ai.lti_registration
+
+    def _get_course_instructor(self, lms_course: LMSCourse) -> LMSUser | None:
+        return self._db.scalars(
+            select(LMSUser)
+            .join(LMSCourseMembership)
+            .join(LTIRole)
+            .where(
+                LMSCourseMembership.lms_course_id == lms_course.id,
+                LTIRole.type == RoleType.INSTRUCTOR,
+                LTIRole.scope == RoleScope.COURSE,
+            )
+            .order_by(LMSCourseMembership.updated.desc())
+        ).first()
+
+    def _get_canvas_sections(
+        self,
+        canvas_api_client: CanvasAPIClient,
+        oauth2_token_service: OAuth2TokenService,
+        lms_course: LMSCourse,
+        with_refresh_token=False,
+    ) -> list[dict]:
+        try:
+            api_sections = canvas_api_client.course_sections(
+                lms_course.lms_api_course_id, with_students=True
+            )
+        except OAuth2TokenError as err:
+            if with_refresh_token and err.refreshable:
+                if not self._refresh_canvas_token(
+                    canvas_api_client, oauth2_token_service
+                ):
+                    LOG.info(
+                        "Failed to fetch sections for course %s, error refreshing token",
+                        lms_course.id,
+                    )
+                    return []
+
+                api_sections = self._get_canvas_sections(
+                    canvas_api_client,
+                    oauth2_token_service,
+                    lms_course,
+                    with_refresh_token=False,
+                )
+            else:
+                LOG.info(
+                    "Failed to fetch sections for course %s, invalid API token",
+                    lms_course.id,
+                )
+                return []
+
+        return api_sections
```

Comment:
Perhaps you can add some early returns to avoid so much nesting here.

```suggestion
        try:
            return canvas_api_client.course_sections(
                lms_course.lms_api_course_id, with_students=True
            )
        except OAuth2TokenError as err:
            if not with_refresh_token and err.refreshable:
                LOG.info(
                    "Failed to fetch sections for course %s, invalid API token",
                    lms_course.id,
                )
                return []
        
            if not self._refresh_canvas_token(
                canvas_api_client, oauth2_token_service
            ):
                LOG.info(
                    "Failed to fetch sections for course %s, error refreshing token",
                    lms_course.id,
                )
                return []
        
            return self._get_canvas_sections(
                canvas_api_client,
                oauth2_token_service,
                lms_course,
                with_refresh_token=False,
            )
```